File: src/integrations/jira.ts
import type { TicketSpec } from '../types.js';

/**
 * Jira Cloud adapter (Phase 4): fetch a ticket via REST API v3 with
 * scoped API-token Basic auth. Mirrors the Linear adapter shape.
 * Auth: `email:apiToken` base64 — the token inherits the user's permissions,
 * so use a dedicated service account with minimal project scope.
 */

export interface JiraCredentials {
  domain: string; // e.g. acme.atlassian.net
  email: string;
  apiToken: string;
}

export const ADF_TO_TEXT = (node: unknown): string => {
  const n = node as { type?: string; text?: string; content?: unknown[] } | null;
  if (!n) return '';
  if (n.type === 'text') return n.text ?? '';
  if (n.type === 'hardBreak') return '\n';
  // block containers (doc, bulletList, ...) separate children by newline;
  // inline containers (paragraph) concatenate their children directly
  const sep = n.type === 'paragraph' ? '' : '\n';
  return (n.content ?? []).map(ADF_TO_TEXT).join(sep);
};

/** Extract `- [ ]` checklist items under an "Acceptance Criteria" heading; fall back to all checklist items. */
export function extractAcceptanceCriteria(markdown: string): string[] {
  const lines = markdown.split('\n');
  const inSection = /#{1,6}\s*(acceptance criteria|ac)\b/i;
  let inAc = false;
  const all: string[] = [];
  const fromAc: string[] = [];
  for (const line of lines) {
    if (inSection.test(line)) {
      inAc = true;
      continue;
    }
    if (/^#{1,6}\s/.test(line)) inAc = false;
    const m = line.match(/^\s*[-*]\s*\[( |x|X)\]\s*(.+)$/);
    if (m) {
      all.push(m[2]!.trim());
      if (inAc) fromAc.push(m[2]!.trim());
    }
  }
  return fromAc.length > 0 ? fromAc : all;
}

interface JiraIssuePayload {
  key: string;
  fields: {
    summary: string;
    description: unknown; // ADF document
    labels?: string[];
  };
}

export function parseJiraIssue(payload: JiraIssuePayload): TicketSpec {
  const description = ADF_TO_TEXT(payload.fields.description);
  return {
    id: payload.key,
    title: payload.fields.summary,
    description,
    labels: payload.fields.labels ?? [],
    acceptanceCriteria: extractAcceptanceCriteria(description),
    trackerInternalId: payload.key,
  };
}

export async function fetchJiraTicket(
  issueKey: string,
  creds: JiraCredentials,
  opts: { retries?: number; sleep?: (ms: number) => Promise<void> } = {},
): Promise<TicketSpec> {
  const retries = opts.retries ?? 3;
  const sleep = opts.sleep ?? ((ms) => new Promise<void>((r) => setTimeout(r, ms)));
  const auth = Buffer.from(`${creds.email}:${creds.apiToken}`).toString('base64');
  const url = `https://${creds.domain}/rest/api/3/issue/${encodeURIComponent(issueKey)}?fields=summary,description,labels`;

  let res: Response | null = null;
  for (let attempt = 0; attempt <= retries; attempt++) {
    res = await fetch(url, {
      headers: { Authorization: `Basic ${auth}`, Accept: 'application/json' },
    });
    if (res.status !== 429 || attempt === retries) break;
    const retryAfter = Number(res.headers.get('retry-after'));
    const delayMs =
      Number.isFinite(retryAfter) && retryAfter > 0
        ? retryAfter * 1000
        : Math.min(30_000, 1000 * 2 ** attempt) + Math.floor(Math.random() * 500);
    await sleep(delayMs);
  }

  if (!res!.ok) {
    throw new Error(`Jira API request failed: HTTP ${res!.status}`);
  }
  const payload = (await res!.json()) as JiraIssuePayload & { errorMessages?: string[] };
  if (!payload?.key) {
    throw new Error(`Jira issue not found: ${issueKey}`);
  }
  return parseJiraIssue(payload);
}

File: internal/integrations/jira.go
// Jira Cloud adapter (Phase 4): the Go port of src/integrations/jira.ts.
// Fetches a ticket via REST API v3 with scoped API-token Basic auth.
// Mirrors the Linear adapter shape. Auth: `email:apiToken` base64 — the
// token inherits the user's permissions, so use a dedicated service account
// with minimal project scope.
package integrations

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
)

// JiraCredentials is the REST v3 auth triple.
type JiraCredentials struct {
	// Domain, e.g. "acme.atlassian.net".
	Domain string
	// Email of the token owner.
	Email string
	// APIToken inherits the user's permissions; use a dedicated service
	// account with minimal project scope.
	APIToken string
}

// ADFToText mirrors ADF_TO_TEXT: flattens an Atlassian Document Format
// document to plain text. Text nodes render their text, hardBreak a
// newline; block containers join children with newlines, inline containers
// (paragraph) concatenate their children directly.
func ADFToText(node any) string {
	n, ok := node.(map[string]any)
	if !ok {
		return ""
	}
	typ, _ := n["type"].(string)
	if typ == "text" {
		if t, ok := n["text"].(string); ok {
			return t
		}
		return ""
	}
	if typ == "hardBreak" {
		return "\n"
	}
	// block containers (doc, bulletList, ...) separate children by newline;
	// inline containers (paragraph) concatenate their children directly
	sep := "\n"
	if typ == "paragraph" {
		sep = ""
	}
	children, _ := n["content"].([]any)
	parts := make([]string, 0, len(children))
	for _, c := range children {
		parts = append(parts, ADFToText(c))
	}
	return strings.Join(parts, sep)
}

var (
	jiraHeadingRe        = regexp.MustCompile(`^#{1,6}\s`)
	jiraChecklistRe      = regexp.MustCompile(`^\s*[-*]\s*\[( |x|X)\]\s*(.+)$`)
	jiraAcceptanceSecsel = regexp.MustCompile(`(?i)#{1,6}\s*(acceptance criteria|ac)\b`)
)

// ExtractAcceptanceCriteria mirrors jira.extractAcceptanceCriteria (also
// used by the GitHub Issues adapter): extract `- [ ]` checklist items under
// an "Acceptance Criteria" heading; fall back to all checklist items.
func ExtractAcceptanceCriteria(markdown string) []string {
	lines := strings.Split(markdown, "\n")
	inAC := false
	all := []string{}
	fromAC := []string{}
	for _, line := range lines {
		if jiraAcceptanceSecsel.MatchString(line) {
			inAC = true
			continue
		}
		if jiraHeadingRe.MatchString(line) {
			inAC = false
		}
		if m := jiraChecklistRe.FindStringSubmatch(line); m != nil {
			item := strings.TrimSpace(m[2])
			all = append(all, item)
			if inAC {
				fromAC = append(fromAC, item)
			}
		}
	}
	if len(fromAC) > 0 {
		return fromAC
	}
	return all
}

// jiraIssuePayload mirrors the REST v3 issue shape this adapter consumes.
type jiraIssuePayload struct {
	Key    string `json:"key"`
	Fields struct {
		Summary     string   `json:"summary"`
		Description any      `json:"description"` // ADF document
		Labels      []string `json:"labels"`
	} `json:"fields"`
}

// ParseJiraIssue mirrors parseJiraIssue.
func ParseJiraIssue(payload jiraIssuePayload) TicketSpec {
	description := ADFToText(payload.Fields.Description)
	labels := payload.Fields.Labels
	if labels == nil {
		labels = []string{}
	}
	return TicketSpec{
		ID:                 payload.Key,
		Title:              payload.Fields.Summary,
		Description:        description,
		Labels:             labels,
		AcceptanceCriteria: ExtractAcceptanceCriteria(description),
		TrackerInternalID:  payload.Key,
	}
}

// FetchJiraTicket mirrors fetchJiraTicket: REST v3 fetch with 429 retries
// honoring Retry-After. Cross-runtime note: a V8 JSON parse failure would
// throw SyntaxError here — Go surfaces the decoder error instead (no
// fixture pins that message).
func FetchJiraTicket(issueKey string, creds JiraCredentials, opts FetchOptions) (TicketSpec, error) {
	auth := base64.StdEncoding.EncodeToString([]byte(creds.Email + ":" + creds.APIToken))
	url := fmt.Sprintf("https://%s/rest/api/3/issue/%s?fields=summary,description,labels",
		creds.Domain, encodeURIComponent(issueKey))
	res, err := fetchWithRetry(func() (*http.Request, error) {
		req, err := http.NewRequest(http.MethodGet, url, nil)
		if err != nil {
			return nil, err
		}
		req.Header.Set("Authorization", "Basic "+auth)
		req.Header.Set("Accept", "application/json")
		return req, nil
	}, opts, func(status int) bool { return status == 429 })
	if err != nil {
		return TicketSpec{}, err
	}
	defer res.Body.Close()
	raw, err := io.ReadAll(res.Body)
	if err != nil {
		return TicketSpec{}, err
	}

	if !isOK(res.StatusCode) {
		return TicketSpec{}, fmt.Errorf("Jira API request failed: HTTP %d", res.StatusCode)
	}

	var payload jiraIssuePayload
	if err := json.Unmarshal(raw, &payload); err != nil {
		return TicketSpec{}, fmt.Errorf("invalid Jira response: %w", err)
	}
	if payload.Key == "" {
		return TicketSpec{}, fmt.Errorf("Jira issue not found: %s", issueKey)
	}
	return ParseJiraIssue(payload), nil
}
